File: packages/protocol/src/rule-evaluator.ts
import { Rule, RuleOp } from './field-map.js';

/**
 * Pure hysteresis/debounce state machine shared by the host's cross-device
 * rule engine and the browser's dry-run simulator. The firmware's local
 * evaluator (C) mirrors this same semantics — see ARCHITECTURE.md §6 — but
 * is implemented separately since it must run with zero dynamic allocation.
 */

export function compareOp(op: RuleOp, value: number, threshold: number): boolean {
  switch (op) {
    case RuleOp.GT:
      return value > threshold;
    case RuleOp.LT:
      return value < threshold;
    case RuleOp.GTE:
      return value >= threshold;
    case RuleOp.LTE:
      return value <= threshold;
    case RuleOp.EQ:
      return value === threshold;
    case RuleOp.NEQ:
      return value !== threshold;
  }
}

/** True once the value has crossed back past the hysteresis band, allowing the rule to re-arm. */
export function isReset(op: RuleOp, value: number, threshold: number, hysteresis: number): boolean {
  switch (op) {
    case RuleOp.GT:
    case RuleOp.GTE:
      return value < threshold - hysteresis;
    case RuleOp.LT:
    case RuleOp.LTE:
      return value > threshold + hysteresis;
    case RuleOp.EQ:
    case RuleOp.NEQ:
      return !compareOp(op, value, threshold);
  }
}

export interface RuleEvaluatorState {
  armed: boolean; // true = condition not yet fired since the last reset
  conditionSinceMs: number | null; // when the condition started holding continuously, for debounce
  fired: boolean; // true once the action has fired for the current arm cycle
}

export function createRuleState(): RuleEvaluatorState {
  return { armed: true, conditionSinceMs: null, fired: false };
}

/**
 * Feeds one new reading through a single rule's state machine. Returns true
 * exactly once per arm cycle, the moment the condition has held continuously
 * for `debounce_ms`. Mutates `state` in place.
 */
export function stepRule(
  state: RuleEvaluatorState,
  rule: Pick<Rule, 'op' | 'threshold' | 'hysteresis' | 'debounce_ms'>,
  value: number,
  atMs: number,
): boolean {
  if (!state.armed) {
    if (isReset(rule.op, value, rule.threshold, rule.hysteresis)) {
      state.armed = true;
      state.fired = false;
      state.conditionSinceMs = null;
    }
    return false;
  }

  const holds = compareOp(rule.op, value, rule.threshold);
  if (!holds) {
    state.conditionSinceMs = null;
    return false;
  }

  if (state.conditionSinceMs === null) {
    state.conditionSinceMs = atMs;
  }

  const held = atMs - state.conditionSinceMs;
  if (!state.fired && held >= rule.debounce_ms) {
    state.fired = true;
    state.armed = false;
    return true;
  }
  return false;
}

File: firmware/common/src/rule_evaluator.c
#include "lorahome/rule_evaluator.h"

static bool compare_op(lorahome_rule_op_t op, float value, float threshold) {
  switch (op) {
    case LORAHOME_OP_GT:
      return value > threshold;
    case LORAHOME_OP_LT:
      return value < threshold;
    case LORAHOME_OP_GTE:
      return value >= threshold;
    case LORAHOME_OP_LTE:
      return value <= threshold;
    case LORAHOME_OP_EQ:
      return value == threshold;
    case LORAHOME_OP_NEQ:
      return value != threshold;
    default:
      return false;
  }
}

/* True once the value has crossed back past the hysteresis band, allowing the rule to re-arm. */
static bool is_reset(lorahome_rule_op_t op, float value, float threshold, float hysteresis) {
  switch (op) {
    case LORAHOME_OP_GT:
    case LORAHOME_OP_GTE:
      return value < threshold - hysteresis;
    case LORAHOME_OP_LT:
    case LORAHOME_OP_LTE:
      return value > threshold + hysteresis;
    case LORAHOME_OP_EQ:
    case LORAHOME_OP_NEQ:
      return !compare_op(op, value, threshold);
    default:
      return true;
  }
}

void lorahome_rule_state_init(lorahome_rule_state_t* state) {
  state->armed = true;
  state->has_condition_since = false;
  state->condition_since_ms = 0;
  state->fired = false;
}

bool lorahome_rule_step(lorahome_rule_state_t* state, const lorahome_rule_t* rule, float value, uint32_t at_ms) {
  const lorahome_rule_op_t op = (lorahome_rule_op_t)rule->op;

  if (!state->armed) {
    if (is_reset(op, value, rule->threshold, rule->hysteresis)) {
      state->armed = true;
      state->fired = false;
      state->has_condition_since = false;
    }
    return false;
  }

  const bool holds = compare_op(op, value, rule->threshold);
  if (!holds) {
    state->has_condition_since = false;
    return false;
  }

  if (!state->has_condition_since) {
    state->condition_since_ms = at_ms;
    state->has_condition_since = true;
  }

  const uint32_t held_ms = at_ms - state->condition_since_ms;
  if (!state->fired && held_ms >= rule->debounce_ms) {
    state->fired = true;
    state->armed = false;
    return true;
  }
  return false;
}
